### ISSUE TITLE
Unable to delete or merge user in Tendenci 14.0.4

### ISSUE THREAD

**Issue by rob-hills** [tendenci/tendenci#1093]
Our organisation rarely needs to delete or merge user records, so we're unsure when this problem first arose, but we are now finding we can't delete a user or merge duplicate user accounts, whereas we have been able to do this in the past.

Here is the traceback from a recent attempt:

```
Traceback (most recent call last):
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)

The above exception (could not load library "/usr/lib/postgresql/11/lib/postgis-2.5.so": /usr/lib/x86_64-linux-gnu/libSFCGAL.so.1: undefined symbol: _ZlsRSoPK12__mpz_struct
) was the direct cause of the following exception:
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/contrib/admin/options.py", line 616, in wrapper
    return self.admin_site.admin_view(view)(*args, **kwargs)
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/utils/decorators.py", line 130, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/views/decorators/cache.py", line 44, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/contrib/admin/sites.py", line 232, in inner
    return view(request, *args, **kwargs)
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/utils/decorators.py", line 43, in _wrapper
    return bound_method(*args, **kwargs)
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/utils/decorators.py", line 130, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/contrib/admin/options.py", line 1723, in changelist_view
    response = self.response_action(request, queryset=cl.get_queryset(request))
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/contrib/admin/options.py", line 1408, in response_action
    response = func(self, request, queryset)
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/tendenci/apps/perms/admin.py", line 146, in soft_delete_selected
    deletable_objects, count, perms_needed, protected = get_deleted_objects(
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/contrib/admin/utils.py", line 119, in get_deleted_objects
    collector.collect(objs)
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/contrib/admin/utils.py", line 182, in collect
    return super().collect(objs, source_attr=source_attr, **kwargs)
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/db/models/deletion.py", line 295, in collect
    if sub_objs:
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/db/models/query.py", line 284, in __bool__
    self._fetch_all()
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/db/models/query.py", line 1324, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/db/models/query.py", line 51, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/db/models/sql/compiler.py", line 1175, in execute_sql
    cursor.execute(sql, params)
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/srv/tendenci/environment.py-3.8/lib/python3.8/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)

Exception Type: InternalError at /admin/auth/user/
Exception Value: could not load library "/usr/lib/postgresql/11/lib/postgis-2.5.so": /usr/lib/x86_64-linux-gnu/libSFCGAL.so.1: undefined symbol: _ZlsRSoPK12__mpz_struct

```

**Comment by rob-hills** [comment]
Preliminary research into this problem took me to [this email exchange](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=947289) which suggests that the problem may be due to a version mismatch between the postgis library installed in the OS and the extension in the database.

Executing:
```
SELECT extversion
FROM pg_catalog.pg_extension
WHERE extname='postgis';
```
within our Test database returned: 
```
 extversion 
------------
 2.5.5
(1 row)
```
Checking the version of Postgis that we have installed in our OS (Ubuntu 20.04), I used `apt-cache policy postgis` which returned:
```
postgis:
  Installed: (none)
  Candidate: 3.2.1+dfsg-1.pgdg18.04+1
  Version table:
     3.2.1+dfsg-1.pgdg18.04+1 500
        500 http://apt.postgresql.org/pub/repos/apt bionic-pgdg/main amd64 Packages
     3.0.0+dfsg-6ubuntu4 500
        500 http://mirror.binarylane.com.au/ubuntu focal/universe amd64 Packages
        500 http://archive.ubuntu.com/ubuntu focal/universe amd64 Packages

```
**HEY!**  Postgis not installed at all?  So we don't have a version mismatch so much as a complete hole where our library should be!

Clearly we had Postgis installed originally as I know you can't install the extension within Postgres without the postgis package having been installed first.  However it appears that somewhere along the way our Postgres database has been upgraded, quietly losing the postgis Ubuntu package in the process.  Thinking on it a bit more, I suspect our recent Ubuntu upgrade from 18.04LTS to 20.04LTS may have been the culprit.  I confess I am surprised the impact on Tendenci functionality has been so subtle!

So it would appear the appropriate fix for this issue will be to reinstall the Ubuntu postgis package and upgrade the postgis extensions within our Tendenci production and test databases.  Is this correct?

**Comment by rob-hills** [comment]
Further research suggests that I can run:
```
sudo apt install postgis
```
in the Ubuntu terminal and then, from the [Postgis docco](https://postgis.net/docs/PostGIS_Extensions_Upgrade.html) run the following in each of our databases:
```
SELECT PostGIS_Extensions_Upgrade();
```

Is this an appropriate way to go?

I know this is a Postgres/Postgis/OS issue, but should there be something about it in the Tendenci ReadTheDocs docco?  I'm happy to contribute if you agree it's worth doing.

**Comment by jennyq** [comment]
Hmm, looks like you'll need to do that. If the postgis extension doesn't exist, you can add it with "CREATE extension postgis;".

**Comment by rob-hills** [comment]
Unfortunately @jennyq the extension does exist.  I've installed the current postgis.  `apt-cache policy postgis` now returns the following:

```
postgis:
  Installed: 3.2.1+dfsg-1.pgdg18.04+1
  Candidate: 3.2.1+dfsg-1.pgdg18.04+1
```
so it appears that I now have the latest available version of postgis installed?  That said, I'm not sure about the pgdg18.04 when we're now running Ubuntu 20.04 here?

When I try and upgrade our test database I get the following error:
```
tendenci_aeva_test=# SELECT PostGIS_Extensions_Upgrade();
ERROR:  could not load library "/usr/lib/postgresql/11/lib/postgis-2.5.so": /usr/lib/x86_64-linux-gnu/libSFCGAL.so.1: undefined symbol: _ZlsRSoPK12__mpz_struct
CONTEXT:  SQL statement "SELECT public.postgis_lib_version()"
PL/pgSQL function postgis_full_version() line 26 at SQL statement
SQL statement "SELECT public.postgis_full_version()"
PL/pgSQL function postgis_extensions_upgrade() line 26 at RETURN
```
So the upgrade process seems to be trying to call the old libraries still.

For the record, I've tried rebooting the server just in case there were some old libraries hanging around in memory, but it hasn't helped.

I thought I'd try dropping the extensions and reinstalling them in our TEST database.  I was able to `drop extension postgis_topology;` and `drop extension postgis_tiger_geocoder;' without any problem, but when I tried `drop extension postgis;' I ran into dependency issues:
```
ERROR:  cannot drop extension postgis because other objects depend on it
DETAIL:  column loc of table social_services_reliefassessment depends on type geometry
column loc of table social_services_skillset depends on type geometry
HINT:  Use DROP ... CASCADE to drop the dependent objects too.

```
I seem to be a bit stuck here!  Do you have any recommendations for the best way through this bind?

**Comment by jennyq** [comment]
Have you tried to stop the tendenci service before dropping the postgis extension?

**Comment by rob-hills** [comment]
Unfortunately stopping the tendenci service, or even tendenci **and** postgresql doesn't help.

I spent more time on this problem today.  I've now established that my problem is caused by a library upgrade that came with the Ubuntu Upgrade that has broken some of the PostGIS functionality.  [This link](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=947289) seems to describe my situation.

My biggest problem is that I can't run pg_dump to backup my production database because it needs an older version (1.2) of the libsfcgal1 library.  However, my OS (Ubuntu 20.04) only provides versions 1.3.5 and 1.3.7.  I've so far not been able to find any way of installing the older version of this library!

I'm not sure how we are supposed to discover hidden gotchas like this one **before** we upgrade, which is obviously the best time to find it out.  Although I had a backup of the DB from before the OS upgrade, by the time we discovered this problem, that backup was way out of date!

**Comment by rob-hills** [comment]
In our Test database (which was a copy of the Production DB made prior to the OS upgrade) I have been able to drop the PostGIS extension by using the CASCADE facility which blatted one column in each of two tables:

```
tendenci_aeva_test=# drop extension postgis;
ERROR:  cannot drop extension postgis because other objects depend on it
DETAIL:  column loc of table social_services_reliefassessment depends on type geometry
column loc of table social_services_skillset depends on type geometry
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
tendenci_aeva_test=# drop extension postgis CASCADE;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to column loc of table social_services_reliefassessment
drop cascades to column loc of table social_services_skillset
DROP EXTENSION

So, do the "loc" columns in social_services_skillset and social_services_reliefassessment contain critical data?  Is it data that could be regenerated somehow?

I'm wondering if I can drop these two columns, backup the rest of my production DB, drop the old PostGIS extension, install the new one and then recreate these two columns?
```

**Comment by rob-hills** [comment]
@jennyq are you able to advise what data we might have in the columns social_services_skillset.loc and social_services_reliefassessment.loc?  Would we lose anything if I drop these columns so I can backup our database and then hopefully upgrade it and get around the block we currently have (see previous two messages above)

**Comment by jennyq** [comment]
@rob-hills, If you're not using the social_services module, you would not lose anything. You can always add those two columns back once you're done with the backup.

**Comment by rob-hills** [comment]
@jennyq , any hints where I might find some good SQL to add the columns back?

I tried looking at the table structures with psql, but of course it bugs out with an error!

I looked through the source code, but of course all DB metadata stuff is done via django.db

**Comment by rob-hills** [comment]
For the record, I have been able to drop the offending columns, drop the geolocation extensions, backup the database, upgrade Postgres (from 11 to 14), reinstall the geolocation extensions and upgrade our Tendenci databases.

This has resolved some issues we had hanging off this, like being unable to recreate our test server (because it needed to copy the production DB which it couldn't because of the broken geolocation stuff) and hence being unable to upgrade Tendenci (we always upgrade our Test server first to ensure we don't get caught out by unanticipated errors).

However, we still can't delete or merge users as we now get an error about the missing social_services_skillset.loc column, so it would be good to be able to put this back and hopefully close this issue at last!

**Comment by jennyq** [comment]
@rob-hills ,

Have you tried these sqls?

    ALTER TABLE social_services_skillset 
    ADD COLUMN loc geometry;

    ALTER TABLE social_services_reliefassessment
    ADD COLUMN loc geometry;

If it still doesn't work, I'd suggest to just drop those two tables, and remove the social_services entry in the django_migrations table, then run `python manage.py migrate social_services`.

**Comment by rob-hills** [comment]
Many thanks for those @jennyq.  For the record, I was able to drop the offending columns, dump a copy of our DB, upgrade Postgres to version 14, upgrading our Tendenci database in the process.  I was then able to add the offending columns back using the SQL you provided above and everything is now working as it should.  Hence, this issue has been resolved and I can close it.
Many thanks for your help!
